# Введение в обработку текста на естественном языке

Материалы:
* Макрушин С.В. Лекция 9: Введение в обработку текста на естественном языке\
* https://realpython.com/nltk-nlp-python/
* https://scikit-learn.org/stable/modules/feature_extraction.html

## Задачи для совместного разбора

In [ ]:
%pip install pymorphy2

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.5/55.5 kB 775.3 kB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.2/8.2 MB 25.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for docopt: filename=docopt-0.6.2-py2.py3-none-any.whl size=13706 sha256=d1fb83604982fb683aaf8b4ea1600ed0d5aa610bad6ccf73552bc0ab97432366
  Stored in directory: /root/.cache/pip/wheels/fc/ab/d4/5da2067ac95b36618c629a5f93f809425700506f72c9732fac
Successfully built docopt


In [ ]:
%pip install strsimpy

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.9/45.9 kB 1.7 MB/s eta 0:00:00


In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
import pymorphy2
import pandas as pd
import re
from strsimpy.longest_common_subsequence import LongestCommonSubsequence

ModuleNotFoundError: No module named 'strsimpy'

1. Считайте слова из файла `litw-win.txt` и запишите их в список `words`. В заданном предложении исправьте все опечатки, заменив слова с опечатками на ближайшие (в смысле расстояния Левенштейна) к ним слова из списка `words`. Считайте, что в слове есть опечатка, если данное слово не содержится в списке `words`.

In [ ]:
text = '''с велечайшим усилием выбравшись из потока убегающих людей Кутузов со свитой уменьшевшейся вдвое поехал на звуки выстрелов русских орудий'''

In [ ]:
def count_lv(w1, w2):
  len1 = len(w1)
  len2 = len(w2)
  res = abs(len1 - len2)
  for i in range(min(len1, len2)):
    if w1[i] != w2[i]:
      res += 2
  return res

In [ ]:
lines = pd.read_csv('litw-win.txt', encoding='cp1251', header=None, delim_whitespace=True)
words = list(lines[1])

sent = text.split()

for wrd in sent:
  if wrd.lower() not in words:
    min_dist = count_lv(wrd, words[0])
    res = words[0]
    for correct_word in words:
      if count_lv(correct_word, wrd) <= min_dist:
        min_dist = count_lv(correct_word, wrd)
        res = correct_word
    sent[sent.index(wrd)] = res
print(' '.join(sent))

с величайшим усилием выбравшись из потока убегающих людей Кутузов со свитой уменьшившейся вдвое поехал на звуки выстрелов русских орудий


2. Разбейте текст из формулировки задания 1 на слова; проведите стемминг и лемматизацию слов.

In [ ]:
text = 'Считайте слова из файла litw-win.txt и запишите их в список words. В заданном предложении исправьте все опечатки, заменив слова с опечатками на ближайшие (в смысле расстояния Левенштейна) к ним слова из списка words. Считайте, что в слове есть опечатка, если данное слово не содержится в списке words.'
def tknz(data):
  return re.findall(r'[а-яА-яЁёa-zA-z]+', data)
res = tknz(text)
res

['Считайте',
 'слова',
 'из',
 'файла',
 'litw',
 'win',
 'txt',
 'и',
 'запишите',
 'их',
 'в',
 'список',
 'words',
 'В',
 'заданном',
 'предложении',
 'исправьте',
 'все',
 'опечатки',
 'заменив',
 'слова',
 'с',
 'опечатками',
 'на',
 'ближайшие',
 'в',
 'смысле',
 'расстояния',
 'Левенштейна',
 'к',
 'ним',
 'слова',
 'из',
 'списка',
 'words',
 'Считайте',
 'что',
 'в',
 'слове',
 'есть',
 'опечатка',
 'если',
 'данное',
 'слово',
 'не',
 'содержится',
 'в',
 'списке',
 'words']

In [ ]:
from nltk.stem.snowball import SnowballStemmer
stemmer = SnowballStemmer("russian")
stemmed_words = [stemmer.stem(word) for word in res]
print(stemmed_words)

['счита', 'слов', 'из', 'файл', 'litw', 'win', 'txt', 'и', 'запиш', 'их', 'в', 'список', 'words', 'в', 'зада', 'предложен', 'исправьт', 'все', 'опечатк', 'замен', 'слов', 'с', 'опечатк', 'на', 'ближайш', 'в', 'смысл', 'расстоян', 'левенштейн', 'к', 'ним', 'слов', 'из', 'списк', 'words', 'счита', 'что', 'в', 'слов', 'ест', 'опечатк', 'есл', 'дан', 'слов', 'не', 'содерж', 'в', 'списк', 'words']


In [ ]:
from pymystem3 import Mystem
def lemmatize_sentence(text):
    lemmas = Mystem().lemmatize(text)
    return "".join(lemmas).strip()
lemmatize_sentence(text)

'считать слово из файл litw-win.txt и записывать они в список words. в заданный предложение исправлять весь опечатка, заменять слово с опечатка на близкий (в смысл расстояние левенштейн) к они слово из список words. считать, что в слово быть опечатка, если данный слово не содержаться в список words.'

3. Преобразуйте предложения из формулировки задания 1 в векторы при помощи `CountVectorizer`.

## Лабораторная работа 9

In [ ]:
import pymorphy2
import pandas as pd
import re
import nltk
import numpy as np

In [ ]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
from nltk import edit_distance

### Расстояние редактирования

1.1 Загрузите предобработанные описания рецептов из файла `preprocessed_descriptions.csv`. Получите набор уникальных слов `words`, содержащихся в текстах описаний рецептов (воспользуйтесь `word_tokenize` из `nltk`).

In [ ]:
data = [str(d) for d in pd.read_csv('preprocessed_descriptions.csv')['preprocessed_descriptions']]
words = list(set(nltk.word_tokenize(' '.join(data))))
words_all = list(nltk.word_tokenize(' '.join(data)))
words[:7]

['motherin',
 'lemons',
 'preschool',
 'bakednot',
 'versionmight',
 'disha',
 'consistencyto']

1.2 Сгенерируйте 5 пар случайно выбранных слов и посчитайте между ними расстояние редактирования.

In [ ]:
random_words = np.random.choice(words, size=5, replace=False)
random_words

array(['stepped', 'imported', 'roberts', 'wonderul', 'pleasing'],
      dtype='<U146')

In [ ]:
for w1 in random_words:
  for w2 in random_words:
    print(f'Для слова {w1} и {w2} расстояние: {edit_distance(w1, w2, substitution_cost=2)}')

Для слова stepped и stepped расстояние: 0
Для слова stepped и imported расстояние: 9
Для слова stepped и roberts расстояние: 12
Для слова stepped и wonderul расстояние: 13
Для слова stepped и pleasing расстояние: 11
Для слова imported и stepped расстояние: 9
Для слова imported и imported расстояние: 0
Для слова imported и roberts расстояние: 9
Для слова imported и wonderul расстояние: 12
Для слова imported и pleasing расстояние: 12
Для слова roberts и stepped расстояние: 12
Для слова roberts и imported расстояние: 9
Для слова roberts и roberts расстояние: 0
Для слова roberts и wonderul расстояние: 9
Для слова roberts и pleasing расстояние: 11
Для слова wonderul и stepped расстояние: 13
Для слова wonderul и imported расстояние: 12
Для слова wonderul и roberts расстояние: 9
Для слова wonderul и wonderul расстояние: 0
Для слова wonderul и pleasing расстояние: 14
Для слова pleasing и stepped расстояние: 11
Для слова pleasing и imported расстояние: 12
Для слова pleasing и roberts расстояние

1.3 Напишите функцию, которая для заданного слова `word` возвращает `k` ближайших к нему слов из списка `words` (близость слов измеряется с помощью расстояния Левенштейна)

In [ ]:
def find1(word, k):
  # словарь вида "слово: расстояние"
  res = {}
  for i in range(k):
    res[words[i]] = edit_distance(words[i], word, substitution_cost=2)
  for w in words[k:]:
    l_dist = edit_distance(w, word)
    for k, v in res.items():
      if v > l_dist:
        del res[k]
        res[w] = l_dist
        break
  return res

In [ ]:
find1('pleasing', 5)

{'placing': 2, 'releasing': 2, 'cleansing': 2, 'pleading': 1, 'pleasing': 0}

### Стемминг, лемматизация

2.1 На основе результатов 1.1 создайте `pd.DataFrame` со столбцами:
    * word
    * stemmed_word
    * normalized_word

Столбец `word` укажите в качестве индекса.

Для стемминга воспользуйтесь `SnowballStemmer`, для нормализации слов - `WordNetLemmatizer`. Сравните результаты стемминга и лемматизации.

In [ ]:
from nltk.stem import SnowballStemmer, WordNetLemmatizer
import pymorphy2

In [ ]:
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...


True

In [ ]:
df = pd.DataFrame()
stemmer_eng = SnowballStemmer('english')
wnl = WordNetLemmatizer()
df['words'] = words
df['stemmed_word'] = [stemmer_eng.stem(i) for i in words]
df['normalized_word'] = [wnl.lemmatize(i) for i in words]
df

words  stemmed_word normalized_word
0          motherin      motherin        motherin
1            lemons         lemon           lemon
2         preschool     preschool       preschool
3          bakednot      bakednot        bakednot
4      versionmight  versionmight    versionmight
...             ...           ...             ...
32863      december        decemb        december
32864         oct11         oct11           oct11
32865          sort          sort            sort
32866   excessively        excess     excessively
32867         grown         grown           grown

[32868 rows x 3 columns]

2.2. Удалите стоп-слова из описаний рецептов. Какую долю об общего количества слов составляли стоп-слова? Сравните топ-10 самых часто употребляемых слов до и после удаления стоп-слов.

In [ ]:
from nltk.probability import FreqDist
# топ-10 самых часто употребляемых слов до удаления стоп-слов
freq_list = FreqDist(words_all)
freq_list.most_common(10)

[('the', 40072),
 ('a', 34951),
 ('and', 30245),
 ('this', 26859),
 ('i', 24836),
 ('to', 23471),
 ('is', 20285),
 ('it', 19756),
 ('of', 18364),
 ('for', 15939)]

In [ ]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [ ]:
from nltk.corpus import stopwords
# топ-10 самых часто употребляемых слов после удаления стоп-слов
stop_words_eng = set(stopwords.words('english'))
words_all_sw = [w for w in words_all if w in stop_words_eng]
words_all_without_sw =[w for w in words_all if w not in stop_words_eng]
freq_list_sp = FreqDist(words_all_sw)
freq_list_without_sp = FreqDist(words_all_without_sw)
freq_list_without_sp.most_common(10)

[('recipe', 14871),
 ('make', 6326),
 ('time', 5137),
 ('use', 4620),
 ('great', 4430),
 ('like', 4167),
 ('easy', 4152),
 ('one', 3872),
 ('made', 3810),
 ('good', 3791)]

In [ ]:
print(f'Доля, которую стоп-слова составляют от общего количества: {freq_list_sp.N() / freq_list.N()}')

Доля, которую стоп-слова составляют от общего количества: 0.4564649471672189


### Векторное представление текста

3.1 Выберите случайным образом 5 рецептов из набора данных. Представьте описание каждого рецепта в виде числового вектора при помощи `TfidfVectorizer`

In [ ]:
rec = pd.read_csv('preprocessed_descriptions.csv')
random_words = np.random.randint(0, rec.shape[0] - 1, size=5)
random_rcps = rec.iloc[random_words]
random_rcps

name  \
4547         candy bar apertif   
18009  mom s old fashion fudge   
27143      tex mex cheese cake   
18799          noodles parmano   
23342           santa fe salsa   

                               preprocessed_descriptions  
4547   tastes just like the chewy chewy candy that la...  
18009  this is so good do not skimp on the ingredient...  
27143  great to bring this smoking cheese cake to par...  
18799  i tried so hard to recreate the taste of the n...  
23342  i love this recipe because its so easy to make...

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
tv = TfidfVectorizer()

In [ ]:
rcps_tv = tv.fit_transform(list(random_rcps['preprocessed_descriptions']))
vectors = rcps_tv.toarray()
vectors[:2]

array([[0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.29265836, 0.        , 0.58531672, 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.29265836, 0.        , 0.29265836,
        0.29265836, 0.29265836, 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.  

3.2 Вычислите близость между каждой парой рецептов, выбранных в задании 3.1, используя косинусное расстояние (`scipy.spatial.distance.cosine`) Результаты оформите в виде таблицы `pd.DataFrame`. В качестве названий строк и столбцов используйте названия рецептов.

In [ ]:
cos_dist = vectors @ vectors.T
cos_dist

array([[1.        , 0.01484329, 0.04349745, 0.04621241, 0.02663618],
       [0.01484329, 1.        , 0.04480229, 0.13360026, 0.1717768 ],
       [0.04349745, 0.04480229, 1.        , 0.11244961, 0.24228169],
       [0.04621241, 0.13360026, 0.11244961, 1.        , 0.14091161],
       [0.02663618, 0.1717768 , 0.24228169, 0.14091161, 1.        ]])

In [ ]:
from scipy.spatial.distance import cosine

res = pd.DataFrame(index=random_rcps['name'], columns=random_rcps['name'])

for i in range(5):
  for j in range(5):
    res.iloc[i, j] = cosine(vectors[i], vectors[j])
res

name                    candy bar apertif mom s old fashion fudge  \
name                                                                
candy bar apertif                       0                0.985157   
mom s old fashion fudge          0.985157                       0   
tex mex cheese cake              0.956503                0.955198   
noodles parmano                  0.953788                  0.8664   
santa fe salsa                   0.973364                0.828223   

name                    tex mex cheese cake noodles parmano santa fe salsa  
name                                                                        
candy bar apertif                  0.956503        0.953788       0.973364  
mom s old fashion fudge            0.955198          0.8664       0.828223  
tex mex cheese cake                       0         0.88755       0.757718  
noodles parmano                     0.88755               0       0.859088  
santa fe salsa                     0.757718        0.859088              0

3.3 Какие рецепты являются наиболее похожими? Прокомментируйте результат (словами).

Наиболее похожими оказались рецепты 'tex mex cheese cake' и 'santa fe salsa', так как у них самое маленькое косинусное расстояние (0.757718). У остальных рецептов косинусное расстояние больше 0.8